### PR TITLE
[Feat] Add Cosmos3-Nano Pipeline Parallel support

### PR DESCRIPTION
**Pull request by MrDongsls**
[Feat] Add Cosmos3-Nano Pipeline Parallel support

## Purpose
Part of #4340 

Cosmos3-Nano pipeline-parallel support.

Environment:

GPU: 2× NVIDIA RTX 5880 48GB + 2x NVIDIA RTX 4080 16GB
torch: 2.11.0+cu130
vLLM: 0.23.0

## Changes

1. PipelineParallelMixin: Support multi-modal outputs
    - All CFG paths now operate element-wise, naturally extending to (video, action, sound) without special-casing
2. Cosmos3-Nano pipeline parallel support
    - Pass hidden_gen and UND K/V cache (first step only) between stages via IntermediateTensors
    - Non-last stages skip CFG combination and condition re-injection; participate in scheduler via a dedicated noise_pred is None branch

## Test Plan
Baseline and PP size=2 tests on 2x`NVIDIA RTX 5880`, PP size=2 + CFGP size=2 tests on 2x`NVIDIA RTX 5880` + 2x`NVIDIA RTX 4080`, using the offcial parameters provided in `recipes/cosmos3/Cosmos3-Nano.md`, assets used in tests are downloaded from official repo: https://huggingface.co/nvidia/Cosmos3-Nano

Baseline server:
```bash
vllm serve nvidia/Cosmos3-Nano --omni --host 0.0.0.0 --port 8000 --no-guardrails --init-timeout 1800
```

PP server:
```bash
vllm serve nvidia/Cosmos3-Nano --omni --host 0.0.0.0 --port 8000 --no-guardrails --init-timeout 1800 --pipeline-parallel-size=2
```

PP+CFGP server:
```bash
vllm serve nvidia/Cosmos3-Nano --omni --host 0.0.0.0 --port 8000 --no-guardrails --init-timeout 1800 --pipeline-parallel-size=2 --cfg-parallel-size=2
```

T2I:
```bash
curl -sS -X POST http://localhost:8000/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "model": "nvidia/Cosmos3-Nano",
    "prompt": "A photorealistic red sports car on a city street at golden hour, cinematic lighting.",
    "negative_prompt": "blurry, distorted, low quality",
    "size": "1024x1024", "n": 1, "response_format": "b64_json",
    "num_inference_steps": 50, "guidance_scale": 7.0, "seed": 42
  }' | python3 -c "import sys,json,base64; open('cosmos3_t2i.png','wb').write(base64.b64decode(json.load(sys.stdin)['data'][0]['b64_json']))"
```

T2V:
```bash
curl -sS -X POST http://localhost:8000/v1/videos \
  -H "Accept: video/mp4" \
  -F "model=nvidia/Cosmos3-Nano" \
  -F "prompt=A robot arm is cleaning a plate in the kitchen" \
  -F "negative_prompt=blurry, distorted, low quality, jittery, deformed" \
  -F "size=1280x720" -F "num_frames=189" -F "fps=24" \
  -F "num_inference_steps=35" -F "guidance_scale=6.0" \
  -F "max_sequence_length=4096" -F "flow_shift=10.0" \
  -F 'extra_params={"use_resolution_template":false,"use_duration_template":false}' \
  -F "seed=123" \
  -o cosmos3_t2v.mp4
```

I2V:
```bash
curl -sS -X POST http://localhost:8000/v1/videos \
  -H "Accept: video/mp4" \
  -F "model=nvidia/Cosmos3-Nano" \
  -F "prompt=The scene comes to life with smooth, natural motion." \
  -F "negative_prompt=blurry, distorted, low quality" \
  -F "size=1280x720" -F "num_frames=189" -F "fps=24" \
  -F "num_inference_steps=35" -F "guidance_scale=6.0" \
  -F "max_sequence_length=4096" -F "flow_shift=10.0" \
  -F 'extra_params={"use_resolution_template":false,"use_duration_template":false}' \
  -F "seed=1111" \
  -F "input_reference=@Cosmos3-Nano/assets/example_i2v_input.jpg;type=image/jpeg" \
  -o cosmos3_i2v.mp4
```

## Test Result

| **Task** | **Baseline (sec)** | **PP2 (sec)** | **PP2+CFGP2 (s)** | **PP2 Speedup** | **PP2+CFGP2 Speedup** |
|----------|-----------------|-------------|-------------------|--------------|-------------------|
| T2I | 64.80 | 47.96 | 47.82 | 1.35× | 1.35× |
| T2V | 928.89 | 713.88 | 564.89 | 1.30× | 1.64× |
| I2V | 927.29 | 710.70 | 565.12 | 1.30× | 1.64× |

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)
D K/V cache (first step only) between stages via IntermediateTensors
    - Non-last stages skip CFG combination and condition re-injection; participate in scheduler via a dedicated noise_pred is None branch

Images (left: baseline, middle: PP size=2, right: PP size=2 + CFG size=2)
| **baseline** | **PP size=2** | **PP size=2 + CFG size=2** |
|:---:|:---:|:---:|
| cosmos3_t2i | cosmos3_t2i_pp2 | cosmos3_t2i_pp2cfg |

### CODE REVIEW

**Comment by MrDongsls** [comment]
Videos：
t2v_pp2.mp4
https://github.com/user-attachments/assets/ce5d8fa9-0963-4328-9ca9-8f59e00c3a45
t2v_pp2cfg.mp4
https://github.com/user-attachments/assets/5828643e-f170-4911-ad7f-01f49f2c0997

i2v_baseline.mp4
https://github.com/user-attachments/assets/7bd00dfa-935f-4845-a9d2-86d484a25fbd
i2v_pp2.mp4
https://github.com/user-attachments/assets/886e7c5e-8796-485e-9b3b-6215abde9d31
i2v_pp2cfg.mp4
https://github.com/user-attachments/assets/25ff96fb-b299-4020-90be-6d980161fb33

**Comment by zhumingjue138** [comment]
please add necessary UT cases

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 1]
```diff
@@ -2181,11 +2238,13 @@ def _step(
                 if raw_action_dim is not None and 0 < raw_action_dim < action_pred.shape[-1]:
                     action_pred[..., raw_action_dim:] = 0
             if action_latents is None and sound_latents is None:
-                latents = step_scheduler.step(video_pred, t, latents, return_dict=False)[0]
+                latents = self.scheduler_step_maybe_with_cfg(video_pred, t, latents, do_true_cfg=False)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: This drops the per-call scheduler passed to `diffuse()`: `scheduler_step_maybe_with_cfg()` falls back to `self.scheduler` unless `per_request_scheduler` is provided. Please pass `per_request_scheduler=scheduler` here and in the packed path below so callers using the `scheduler` argument keep the requested scheduler state.

Comment (MrDongsls) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py]: My miss, already fixed. Thanks!

File: vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py [hunk 2]
```diff
@@ -2163,13 +2191,45 @@ def _split_noise_pred(
...
         ) -> torch.Tensor | tuple[torch.Tensor, ...]:
+            if noise_pred is None:
+                # Non-last PP rank: still need to participate in scheduler PP handshake
+                next_latents = self.scheduler_step_maybe_with_cfg(
+                    None, t, latents, do_true_cfg=False, per_request_scheduler=step_scheduler
+                )
+                if _has_multimodal:
+                    if isinstance(next_latents, AsyncLatents):
+                        next_latents = next_latents._resolve()
+                    unpacked = _unpack_joint(next_latents, _joint_shapes, _joint_numels)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:2215]: This breaks PP world_size > 2 when action/sound latents are present. Only rank 0 receives the packed scheduler result from the last stage; middle ranks return their local video latents here, so this unpack treats a video tensor as the packed joint tensor. Gate the unpack to the first PP stage, or keep action/sound unchanged on middle ranks.

Comment (MrDongsls) [reply vllm_omni/diffusion/models/cosmos3/pipeline_cosmos3.py:2215]: Fixed by gating the unpack to is_pipeline_first_stage().

File: vllm_omni/diffusion/distributed/pipeline_parallel.py [hunk 3]
```diff
@@ -17,6 +17,21 @@
 )
 
 
+def _wrap(pred: torch.Tensor | tuple[torch.Tensor, ...]) -> tuple[torch.Tensor, ...]:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/distributed/pipeline_parallel.py:20]: why we have to introduce such util function?

Comment (hsliuustc0106) [reply vllm_omni/diffusion/distributed/pipeline_parallel.py:20]: hard to maintain

**Review by hsliuustc0106 — COMMENTED** [review]
@zzhang-fr PTAL

File: vllm_omni/diffusion/distributed/pipeline_parallel.py (hunk 3 above)
Comment (MrDongsls) [reply vllm_omni/diffusion/distributed/pipeline_parallel.py:20]: I think these are useful helpers for multimodal data transfer across ranks and make pipeline-side code easier to maintain. Maybe better moving them into `vllm_omni/diffusion/distributed/utils.py` for better reusability or remove them entirely and have the model side always return a consistent type?
